format/common/text.go

```diff
@@ -0,0 +1,263 @@
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_IP,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+		FORMAT_TYPE_INTEGER,
+	}
+	RenderExtras = []string{
+		"EtypeName",
+		"ProtoName",
+		"IcmpName",
+	}
+	RenderExtraCall = []RenderExtraFunction{
+		RenderExtraFunctionEtypeName,
+		RenderExtraFunctionProtoName,
+		RenderExtraFunctionIcmpName,
+	}
+)
+
+func AddTextField(name string, jtype int) {
+	TextFields = append(TextFields, name)
+	TextFieldsTypes = append(TextFieldsTypes, jtype)
+}
+
+type RenderExtraFunction func(proto.Message) string
+
+func RenderExtraFetchNumbers(msg proto.Message, fields []string) []uint64 {
+	vfm := reflect.ValueOf(msg)
+	vfm = reflect.Indirect(vfm)
+
+	values := make([]uint64, len(fields))
+	for i, kf := range fields {
+		fieldValue := vfm.FieldByName(kf)
+		if fieldValue.IsValid() {
+			values[i] = fieldValue.Uint()
+		}
+	}
+
+	return values
+}
+
+func RenderExtraFunctionEtypeName(msg proto.Message) string {
+	num := RenderExtraFetchNumbers(msg, []string{"Etype"})
+	return EtypeName[uint32(num[0])]
+}
+
+func RenderExtraFunctionProtoName(msg proto.Message) string {
+	num := RenderExtraFetchNumbers(msg, []string{"Proto"})
+	return ProtoName[uint32(num[0])]
+}
+func RenderExtraFunctionIcmpName(msg proto.Message) string {
+	num := RenderExtraFetchNumbers(msg, []string{"Proto", "IcmpCode", "IcmpType"})
+	return IcmpCodeType(uint32(num[0]), uint32(num[1]), uint32(num[2]))
+}
+
+func IcmpCodeType(proto, icmpCode, icmpType uint32) string {
+	if proto == 1 {
+		return IcmpTypeName[icmpType]
+	} else if proto == 58 {
+		return Icmp6TypeName[icmpType]
+	}
+	return ""
+}
+
+func RenderIP(addr []byte) string {
+	if addr == nil || (len(addr) != 4 && len(addr) != 16) {
+		return ""
+	}
+
+	return net.IP(addr).String()
+}
+
+func FormatMessageReflectText(msg proto.Message, ext string) string {
+	return FormatMessageReflectCustom(msg, ext, "", " ", "=", false)
+}
+
+func FormatMessageReflectJSON(msg proto.Message, ext string) string {
+	return fmt.Sprintf("{%s}", FormatMessageReflectCustom(msg, ext, "\"", ",", ":", true))
+}
+
+func FormatMessageReflectCustom(msg proto.Message, ext, quotes, sep, sign string, null bool) string {
+	fstr := make([]string, len(TextFields)+len(RenderExtras))
+
+	vfm := reflect.ValueOf(msg)
+	vfm = reflect.Indirect(vfm)
+
+	var i int
+	for j, kf := range TextFields {
+		fieldValue := vfm.FieldByName(kf)
+		if fieldValue.IsValid() {
+
+			switch TextFieldsTypes[j] {
+			case FORMAT_TYPE_STRING_FUNC:
+				strMethod := fieldValue.MethodByName("String").Call([]reflect.Value{})
+				fstr[i] = fmt.Sprintf("%s%s%s%s%s%s%s", quotes, kf, quotes, sign, quotes, strMethod[0].String(), quotes)
```

Comment:
If the quote is `"`, if `strMethod[0].String()` contains a `"`, this will not be escaped and generate likely invalid output. (ie. json)

Reply:
I cannot find a nice solution to that :/
It does not only escape the strings, it also adds quotes. This means every string *must* be quoted even in regular `text` format.

Reply:
This is why the json format should really use `encoding/json` and not try to hand-roll it.

Reply:
I agree, but flexibility with the structure required to use `reflect` and preferred to print directly from there. The alternative would have been to create a `map[string]interface{}` which would be iterated again.

Reply:
Slightly ugly, but how about this:
```go
s := strings.TrimPrefix(strings.TrimSuffix(strconv.Quote(string), `"`), `"`)
```

Reply:
If the quotes are always added, I can probably cut the slice directly, no?
I wish the function `quoteWith` (from [strconv](https://golang.org/src/strconv/quote.go?#L19)) was not private.

I'm tempted to leave it as-is.